[1.16.3] Adds overloads of the compat methods to allow for two mod id's

EltrutLikes opened the pull request:
This PR adds overloads of the compat methods to allow for compatibility between two different mods. This is used, for example, in Enhanced Mushrooms for glowshroom beehives or in Cookielicious for sweet berry cookie tile vertical slabs. A common use of these methods is for allowing compatibility between some mod and Quark's vertical slabs and is used often enough that I think it warrants a PR here.

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/BlockSubRegistryHelper.java:
```diff
@@ -236,11 +237,19 @@ public BlockSubRegistryHelper(RegistryHelper parent, DeferredRegister<Item> item
 		return block;
 	}
+	
+	public <B extends Block> RegistryObject<B> createCompatBlock(String modId, String modId2, String name, Supplier<? extends B> supplier, @Nullable ItemGroup group) {
```
This method has no documentation.

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/BlockSubRegistryHelper.java:
```diff
@@ -226,6 +226,7 @@ public BlockSubRegistryHelper(RegistryHelper parent, DeferredRegister<Item> item
 	 * Creates and registers a compat {@link Block}
 	 *
 	 * @param modId    - The mod id of the mod this block is compatible for, set to "indev" for dev tests
+	 * @param modId2   - The mod id of the second mod this block is compatible for, set to "indev" for dev tests (optional)
```
Why is this here, `modId2` isn't even a parameter in this method?

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/BlockSubRegistryHelper.java:
```diff
@@ -236,11 +237,19 @@ public BlockSubRegistryHelper(RegistryHelper parent, DeferredRegister<Item> item
 	 *
-	 * @param modId    - The modId of the mod this block is compatible for, set to "indev" for dev tests
+	 * @param modId    - The mod id of the mod this block is compatible for, set to "indev" for dev tests
+	 * @param modId2   - The mod id of the second mod this block is compatible for, set to "indev" for dev tests (optional)
```
Why is `modId2` here, it isn't even a parameter in this method?

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/BlockSubRegistryHelper.java:
```diff
@@ -252,6 +261,13 @@ public BlockSubRegistryHelper(RegistryHelper parent, DeferredRegister<Item> item
 		return block;
 	}
+	
+	public <B extends Block> RegistryObject<B> createCompatFuelBlock(String modId, String modId2, String name, Supplier<? extends B> supplier, int burnTime, @Nullable ItemGroup group) {
```
This method has no documentation.

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/ItemSubRegistryHelper.java:
```diff
@@ -53,6 +53,7 @@ public ItemSubRegistryHelper(RegistryHelper parent) {
 	 * Creates and registers a compat {@link Item}
 	 *
 	 * @param modId      - The mod id of the mod this item is compatible for, set to "indev" for dev tests
+	 * @param modId2     - The mod id of the second mod this item is compatible for, set to "indev" for dev tests (optional)
```
Why is `modId2` here, it isn't even a parameter in this method?

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/ItemSubRegistryHelper.java:
```diff
@@ -62,6 +63,12 @@ public ItemSubRegistryHelper(RegistryHelper parent) {
 		return item;
 	}
+	
+	public RegistryObject<Item> createCompatItem(String modId, String modId2, String name, Item.Properties properties, ItemGroup group) {
```
This method has no documentation.

SmellyModder reviewed (changes requested):
I like this, but I think it could be done better.
Having separate methods for double-compat for certain methods is kinda redundant since at that point shouldn't it just be checking if a set of mods is loaded rather than having methods for just double-compat? It'd kill an infinite amount of birds with one stone 😛 
Also, it's kinda weird that it checks for both mod ids being indev, maybe just add a boolean as a parameter instead?
There are also some other issues that I've pointed out below.

EltrutLikes commented:
I've fixed the documentation issues and changed the methods to take in an ArrayList of mod id's and a boolean `isIndev` instead of two mod id's. I also added a compat method for the chests as well just for completeness and I switched the parameters of the original chest compat method to match the other compat methods. Let me know if you would like me to revert that though. :)

JacksonPlayz commented:
Would probably be best to use a String varargs instead of an ArrayList, since you don't need to keep the order intact plus it'll just be cleaner for developers to put in multiple strings.

SmellyModder reviewed (changes requested):
As a follow up to what Jackson said, you really shouldn't be making it use `ArrayList` as the compatible mods collection. In the case where you're just iterating over something, it should take in a `Collection` or a vararg, not an `ArrayList` since you're not making use of any of `ArrayList`'s features. 
And in this case, you should use a vararg as it's more friendly when using this method, wouldn't have to initialize a `Collection` object, and will make porting from 1.16.1 to 1.16.3+ easier than having to make a new `ArrayList` for what mods it's compatible for.
Another thing, I think that the `isIndev` parameter should be removed and just check if it's a development environment instead.

EltrutLikes commented:
Apologies for the noobiness, I'm still learning Java.

A question though: if I switch it over to varargs, would you like me to combine the multi-mod compat method, one-mod compat method, and normal method into one (since one-mod and normal would correspond to 1 and 0 parameters, respectively) or just leave them as is?

EltrutLikes commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/BlockSubRegistryHelper.java:
```diff
@@ -1,5 +1,11 @@
 package com.teamabnormals.abnormals_core.core.util.registry;
 
+import java.util.ArrayList;
```
Sorry about that, Eclipse did it automatically when I asked it to fix imports and I didn't notice it until now. I'll fix it in the next commit.

JacksonPlayz commented:
Yes merge them together.

EltrutLikes commented:
I switched over to a String varargs for the mod id list, removed `isIndev` as a parameter in favor of just checking whether it is a dev environment or not, moved the imports back to where they were, and combined the multi-mod and single mod compat methods.

I feel that it is now redundant to have for example both a `createBlock` and `createCompatBlock` method with the String varargs since `createBlock` just corresponds to the `createCompatBlock` with no parameters, but I understand if for clarity you would like to keep both.

bageldotjpg commented:
Ok, sorry for changing up things yet again but Jackson, Smelly and I were discussing in VC yesterday and determined some things.

- It shouldn't automatically make it show up if it is a dev environment, because sometimes that could be an issue; eg: you messed up your compat method so it shows up when the wrong mod is present or it shows up when your own mod is present. They should only show up if you changed something specific so you know you are doing something right.

- To do this, you can just change the modid arg to whatever mod you are working on's modid to test. Because in reality they don't need to be tested that often. e.g: testing a Quark compat object for Atmospheric? Change the compat id to Atmospheric.


We also determined the change to var args, while easy to use, is rather unclean in a development format, and forcing it upon all methods is not a great idea, because that creates a massive breaking change that is also not very readable. So the vararg methods should be new methods and leave the old ones alone.

EltrutLikes commented:
I restored the original compat methods and removed the `if (FMLEnvironment.production)` conditional from the multiple compat methods (which are now once again overloads of the original methods). Hopefully third time is a charm :)

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/BlockSubRegistryHelper.java:
```diff
@@ -11,6 +11,7 @@
 import com.teamabnormals.abnormals_core.common.items.InjectedBlockItem;
 import com.teamabnormals.abnormals_core.common.tileentity.AbnormalsChestTileEntity;
 import com.teamabnormals.abnormals_core.common.tileentity.AbnormalsTrappedChestTileEntity;
+
```
There's no need for this space to be here.

SmellyModder commented on src/main/java/com/teamabnormals/abnormals_core/core/util/registry/ItemSubRegistryHelper.java:
```diff
@@ -5,6 +5,7 @@
 import com.teamabnormals.abnormals_core.common.items.AbnormalsSpawnEggItem;
 import com.teamabnormals.abnormals_core.common.items.FuelItem;
 import com.teamabnormals.abnormals_core.core.registry.BoatRegistry;
+
```
This space doesn't need to be here.

SmellyModder reviewed (changes requested):
There are still a few issues. I've pointed out two minor nitpicks with some imports, and the rest of the problems are with each of the new compatibility methods so, I'll talk about them here.

The varargs parameter at the end of all these new methods is formatted weirdly. 
There is space between String and the dots, and there's no space between the dots and the parameter name.

In my opinion, `modIdList` is a misleading name for the parameter since it's not a list of mod ids, as it's just an array of mod ids. So just naming it `modIds` instead would be more fitting.

The most glaring issue here is the code for determining if the tab should be null. 
It'd be cleaner to have that code for checking if the mods are loaded to be a method somewhere, as it would remove some boilerplate code in these new methods and make it easier for people to add new compatibility methods. Also, I think it'd be better to have `ModList.get()` be a local variable outside the loop.

SmellyModder commented:
I updated the 1.16.3 branch to 1.16.4 so this got closed.
Open a new PR for this on the 1.16.4 branch.
